From c0dc6003d9cf47271465e0c8468eec10059aeb12 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Wed, 27 Nov 2024 11:36:25 +0100
Subject: [PATCH 745/841] feat(st-crypto): manage RNG customization with device
 tree

Implement an optional way of RNG configuration via the device tree which
will override default hard-coded one.
It concerns both NIST and health test configurations.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git bc8e8ae90ccd44120918dc122b5058d8af9fd32c feat(st-crypto): manage RNG customization with device tree]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: Ie4525392b889784d6d4feb655b74b96a4c849257
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/421391
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 drivers/st/crypto/stm32_rng.c | 46 ++++++++++++++++++++++++++++++++---
 1 file changed, 43 insertions(+), 3 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index 04bf7e6f2..407eefb75 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -24,6 +24,7 @@
 #endif
 #if STM32_RNG_VER == 4
 #define DT_RNG_COMPAT			"st,stm32mp13-rng"
+#define DT_RNG_MAX_NIST_CONFIG		3U
 #endif
 #define RNG_CR				0x00U
 #define RNG_SR				0x04U
@@ -36,6 +37,7 @@
 #define RNG_CR_CED			BIT_32(5)
 #if STM32_RNG_VER == 4
 #define RNG_CR_RNG_CONFIG3_SHIFT	8U
+#define RNG_CR_NISTC			BIT_32(12)
 #define RNG_CR_RNG_CONFIG2_SHIFT	13U
 #define RNG_CR_CLKDIV			GENMASK_32(19, 16)
 #define RNG_CR_CLKDIV_SHIFT		16U
@@ -73,6 +75,10 @@
 struct stm32_rng_instance {
 	uintptr_t base;
 	unsigned long clock;
+#if STM32_RNG_VER == 4
+	uint32_t ht_cfg;
+	uint32_t nist_cfg;
+#endif
 };
 
 static struct stm32_rng_instance stm32_rng;
@@ -136,13 +142,12 @@ static int stm32_rng_enable(void)
 
 	/* Update configuration fields */
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_NIST_CONFIG_MASK,
-			   RNG_NIST_CONFIG(RNG_NIST_CONFIG1, RNG_NIST_CONFIG2, RNG_NIST_CONFIG3) |
-			   RNG_CR_CONDRST | RNG_CR_CED);
+			   stm32_rng.nist_cfg | RNG_CR_CONDRST | RNG_CR_CED);
 
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CLKDIV,
 			   (clock_div << RNG_CR_CLKDIV_SHIFT));
 
-	mmio_write_32(stm32_rng.base + RNG_HTCR, RNG_HTCFG_CONFIG);
+	mmio_write_32(stm32_rng.base + RNG_HTCR, stm32_rng.ht_cfg);
 
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CONDRST, RNG_CR_RNGEN);
 #endif
@@ -194,6 +199,39 @@ static int check_data_validity(void)
 	return 0;
 }
 
+static void parse_dt_optional_config(const void *fdt, int node)
+{
+#if STM32_RNG_VER == 4
+	const fdt32_t *cuint;
+	int len;
+	uint32_t nist_dt_config[DT_RNG_MAX_NIST_CONFIG] = {
+		RNG_NIST_CONFIG1,
+		RNG_NIST_CONFIG2,
+		RNG_NIST_CONFIG3
+	};
+
+	cuint = fdt_getprop(fdt, node, "st,rng-cfg", &len);
+	if ((cuint != NULL) && (len > 0) &&
+	    ((uint32_t)len <= (DT_RNG_MAX_NIST_CONFIG * sizeof(uint32_t)))) {
+		uint32_t i;
+
+		for (i = 0U; i < ((uint32_t)len / sizeof(uint32_t)); i++) {
+			nist_dt_config[i] = fdt32_to_cpu(*cuint);
+			cuint++;
+		}
+	}
+
+	stm32_rng.nist_cfg = RNG_NIST_CONFIG(nist_dt_config[0], nist_dt_config[1],
+					     nist_dt_config[2]);
+
+	if (fdt_getprop(fdt, node, "st,rng-cfg-nist-custom", NULL) != NULL) {
+		stm32_rng.nist_cfg |= RNG_CR_NISTC;
+	}
+
+	stm32_rng.ht_cfg = fdt_read_uint32_default(fdt, node, "st,rng-htcfg", RNG_HTCFG_CONFIG);
+#endif
+}
+
 /*
  * stm32_rng_read - Read a number of random bytes from RNG
  * out: pointer to the output buffer
@@ -339,6 +377,8 @@ int stm32_rng_init(void)
 			}
 		}
 
+		parse_dt_optional_config(fdt, node);
+
 		ret = stm32_rng_enable();
 		if (ret != 0) {
 			ERROR("Failed to enable rng@%x\n", dt_rng.base);
-- 
2.47.2

